adi/ad5592r.py

```diff
@@ -0,0 +1,132 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+# RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+# BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from decimal import Decimal
+
+from adi.attribute import attribute
+from adi.context_manager import context_manager
+
+
+class ad5592r(context_manager):
+    """AD5592R ADC/DAC"""
+
+    _complex_data = False
+    channel = []  # type: ignore
+    _device_name = ""
+
+    def __init__(self, uri="", device_name=""):
+
+        context_manager.__init__(self, uri, self._device_name)
+
+        compatible_parts = [
+            "ad5592r",
+            "ad5593r",
+        ]
+
+        self.ctrl = None
+
+        if not device_name:
+            device_name = compatible_parts[0]
+        else:
+            if device_name not in compatible_parts:
+                raise Exception("Not a compatible device: " + device_name)
+
+        # Selecting the device matching device_name AD559XR family as working device.
+        for device in self._ctx.devices:
+            if device.name in device_name:
+                self._ctrl = device
+                break
+
+        # Dynamically get channels after the index
+        for ch in self._ctrl.channels:
+            name = ch._id
+            output = ch._output
+            if name == "temp":
+                setattr(self, name, self._channel(self._ctrl, name, output))
+            else:
+                if output is True:
+                    setattr(
+                        self, name + "_dac", self._channel(self._ctrl, name, output)
+                    )
+                else:
+                    setattr(
+                        self, name + "_adc", self._channel(self._ctrl, name, output)
+                    )
+
+    class _channel(attribute):
+        """AD5592R Input/Output Voltage Channels"""
+
+        # AD559XR voltage channel
+        def __init__(self, ctrl, channel_name, output):
+            self.name = channel_name
+            self._ctrl = ctrl
+            self._output = output
+
+        @property
+        # AD559XR channel raw value
+        def raw(self):
+            return self._get_iio_attr(self.name, "raw", self._output)
+
+        @property
+        # AD559XR channel scale (gain)
+        def scale(self):
+            return float(self._get_iio_attr_str(self.name, "scale", self._output))
+
+        @property
+        # AD559XR channel temp offset value
+        def offset(self):
+            return self._get_iio_attr(self.name, "offset", self._output)
+
+        @raw.setter
+        def raw(self, value):
+            if self._output is True:
+                self._set_iio_attr(self.name, "raw", self._output, value)
```

Comment:
can you add an exception here if its not an output

Reply:
I set up separate channel types for ADC, DAC, and temperature. Adding an exception worked, but it seemed odd to have a setter there in the first place. With the new scheme, setting and getting DAC raw is as expected, attempting to write to an ADC raw gives:
AttributeError: can't set attribute

Also I realized that scale_avaliable doesn't apply to temperature, and offset doesn't apply to ADC or DAC, more reason for separate channel types.

Does this look reasonable?